### ISSUE TITLE
Static libraries and FindOpenVDB.cmake

### ISSUE THREAD

**Issue by joshbainbridge** [AcademySoftwareFoundation/openvdb#521]
What is the recommended procedure to get a static library (.a) from this find module package on a linux system? Looking at the files it appears that setting OPENVDB_USE_STATIC_LIBS would be the correct option. However this is not mentioned in the documentation, and in doing so cmake fails as the static library is passed to get_prerequisites() (calls /usr/bin/ldd), which I've understood to only work with shared libraries.

Before writing a new module for our requirements, just wanted to check if I'm missing something, or why the project does not generate a cmake config package?

**Comment by Idclip** [comment]
Hi @joshbainbridge!

Your assessment is correct, that variable is designed to prioritise finding the static library but there's probably a more CMake-y way to do it. There is indeed an issue with calling `get_prerequisites`, you should be able to resolve this by moving that code into a conditional branch based on the value of `OPENVDB_USE_STATIC_LIBS`:

<details><summary>Example Changes:</summary>
<p>

```
# As the way we resolve optional libraries relies on library file names, use
# the configuration options from the main CMakeLists.txt to allow users
# to manually identify the requirements of OpenVDB builds if they know them.

set(OpenVDB_USES_BLOSC ${USE_BLOSC})
set(OpenVDB_USES_LOG4CPLUS ${USE_LOG4CPLUS})
set(OpenVDB_USES_EXR ${USE_EXR})

if(NOT OPENVDB_USE_STATIC_LIBS)

  # Use GetPrerequisites to see which libraries this OpenVDB lib has linked to
  # which we can query for optional deps. This basically runs ldd/otoll/objdump
  # etc to track deps. We could use a vdb_config binary tools here to improve
  # this process

  include(GetPrerequisites)

  set(_EXCLUDE_SYSTEM_PREREQUISITES 1)
  set(_RECURSE_PREREQUISITES 0)
  set(_OPENVDB_PREREQUISITE_LIST)

  get_prerequisites(${OpenVDB_openvdb_LIBRARY}
    _OPENVDB_PREREQUISITE_LIST
    ${_EXCLUDE_SYSTEM_PREREQUISITES}
    ${_RECURSE_PREREQUISITES}
    ""
    "${SYSTEM_LIBRARY_PATHS}"
  )

  unset(_EXCLUDE_SYSTEM_PREREQUISITES)
  unset(_RECURSE_PREREQUISITES)

  # Search for optional dependencies

  foreach(PREREQUISITE ${_OPENVDB_PREREQUISITE_LIST})
    set(_HAS_DEP)
    get_filename_component(PREREQUISITE ${PREREQUISITE} NAME)

    string(FIND ${PREREQUISITE} "blosc" _HAS_DEP)
    if(NOT ${_HAS_DEP} EQUAL -1)
      set(OpenVDB_USES_BLOSC ON)
    endif()

    string(FIND ${PREREQUISITE} "log4cplus" _HAS_DEP)
    if(NOT ${_HAS_DEP} EQUAL -1)
      set(OpenVDB_USES_LOG4CPLUS ON)
    endif()

    string(FIND ${PREREQUISITE} "IlmImf" _HAS_DEP)
    if(NOT ${_HAS_DEP} EQUAL -1)
      set(OpenVDB_USES_EXR ON)
    endif()
  endforeach()

  unset(_OPENVDB_PREREQUISITE_LIST)
  unset(_HAS_DEP)
endif()
```

</p>
</details>

I'll try and patch this soon. There is a lot of work still required on this particular module, including testing through our CI, support for finding the python and openvdb_houdini libraries and, as you mentioned, introducing a config package. In the build docs we have:

> Note: Typically, projects provide a Config-file rather than a Find-module for downstream use. OpenVDB currently provides a Find-module. This may change in the future. Further information here.

I've not put one of these Config files together before so am not completely sure yet on the form it would take - if you do happen write your own to achieve this, anything you can contribute back would be great.

**Comment by joshbainbridge** [comment]
Thanks Nick,

Cheers for getting back to me so promptly. For the time being I've added a patch to our local version until the changes are added to the main trunk. Will let you know if we have any issues.

I would say that using a variable such as OPENVDB_USE_STATIC_LIBS is fairly standard CMake practice, but it could be listed under the Hints section at the top of the page.

The way this find module package is written, with it being in a relative directory to the binary and header files, is how config packages are designed to work. So you might find using one of these more natural, or at least a nice alternative.

I can put a config package together at some point once there is a good opportunity, and will contribute it back if we do. But the patch you suggested should work for the time being.

**Comment by joshbainbridge** [comment]
Quick update on this, we managed to get things to work, but the symbols for Blosc were not being found when linking the final binary, that is just with the suggested patch. And our fix was to simply link Blosc separately.

However the root issue is most likely due to the Blosc dependency being added under the property IMPORTED_LINK_DEPENDENT_LIBRARIES, as this assumes we are dealing with a shared library...

https://cmake.org/cmake/help/latest/prop_tgt/IMPORTED_LINK_DEPENDENT_LIBRARIES.html

In the case that it is a static build, you probably would want to add the dependency using the standard INTERFACE_LINK_LIBRARIES instead.

**Comment by Idclip** [comment]
@joshbainbridge this should now be resolved (as of #708). Closing, but please re-open if it continued to be an issue
